src/wifi.rs

```diff
@@ -2244,3 +2422,103 @@ where
         AsyncWifi::is_up(self)
     }
 }
+
+pub struct WpsConfig<'a> {
+    pub wps_type: WpsType,
+    pub factory_info: WpsFactoryInfo<'a>,
+}
+
+impl TryFrom<&WpsConfig<'_>> for Newtype<esp_wps_config_t> {
+    type Error = EspError;
+
+    fn try_from(config: &WpsConfig<'_>) -> Result<Self, Self::Error> {
+        let factory_info = Newtype::<wps_factory_information_t>::try_from(&config.factory_info)?.0;
+        Ok(Newtype(esp_wps_config_t {
+            wps_type: config.wps_type.as_raw_type(),
+            factory_info,
+            #[cfg(not(esp_idf_version_major = "4"))]
+            pin: config.wps_type.as_pin(),
+        }))
+    }
+}
+
+pub struct WpsFactoryInfo<'a> {
+    pub manufacturer: &'a str,
+    pub model_number: &'a str,
+    pub model_name: &'a str,
+    pub device_name: &'a str,
+}
+
+impl TryFrom<&WpsFactoryInfo<'_>> for Newtype<wps_factory_information_t> {
+    type Error = EspError;
+
+    fn try_from(info: &WpsFactoryInfo<'_>) -> Result<Self, Self::Error> {
+        let mut result = Newtype(wps_factory_information_t {
+            manufacturer: [0; 65],
+            model_number: [0; 33],
+            model_name: [0; 33],
+            device_name: [0; 33],
+        });
+
+        set_cchar_slice(&mut result.0.manufacturer, info.manufacturer)?;
+        set_cchar_slice(&mut result.0.model_number, info.model_number)?;
+        set_cchar_slice(&mut result.0.model_name, info.model_name)?;
+        set_cchar_slice(&mut result.0.device_name, info.device_name)?;
+
+        Ok(result)
+    }
+}
+
+pub enum WpsType {
+    Pbc,
+    Pin(heapless::String<8>),
+}
+
+impl WpsType {
+    fn as_raw_type(&self) -> wps_type_t {
+        match self {
+            WpsType::Pbc => wps_type_WPS_TYPE_PBC,
+            WpsType::Pin(_) => wps_type_WPS_TYPE_PIN,
+        }
+    }
+
+    #[cfg(not(esp_idf_version_major = "4"))]
+    fn as_pin(&self) -> [ffi::c_char; 9] {
+        match self {
+            WpsType::Pbc => [0; 9],
+            WpsType::Pin(pin) => {
+                let mut result = [0; 9];
+                let bytes = pin.as_bytes();
+                let bytes_cchar = bytes.as_ptr() as *const ffi::c_char;
+                let slice = unsafe { core::slice::from_raw_parts(bytes_cchar, pin.len()) };
+                result[..pin.len()].copy_from_slice(slice);
+                result
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum WpsEvent {
+    Active,
+    Success(Option<Arc<heapless::Vec<WpsCredentials, MAX_WPS_AP_CRED_USIZE>>>),
```

Comment:
Without the `Arc`, clippy will obviously complain because the enum variant will be huge compared to the other ones:

```
warning: large size difference between variants
    --> src\wifi.rs:1777:1
     |
1777 | / pub enum WifiEvent {
1778 | |     Ready,
1779 | |
1780 | |     ScanDone,
...    |
1789 | |     StaWpsSuccess(heapless::Vec<WpsCredentials, MAX_WPS_AP_CRED_USIZE>),
     | |     ------------------------------------------------------------------- the largest variant contains at least 316 bytes
...    |
1792 | |     StaWpsPin(Option<u32>),
     | |     ---------------------- the second-largest variant contains at least 8 bytes
...    |
1803 | |     RocDone,
1804 | | }
     | |_^ the entire enum is at least 320 bytes
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
     = note: `#[warn(clippy::large_enum_variant)]` on by default
help: consider boxing the large fields to reduce the total size of the enum
     |
1789 |     StaWpsSuccess(Box<heapless::Vec<WpsCredentials, MAX_WPS_AP_CRED_USIZE>>),
     |                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The reason I chose `Arc` over `Box` is because it avoids copying the entire vector when the `WifiEvent` is converted to the `WpsStatus`.

Reply:
Also, I decided to split the `WpsStatus::Success` variant up into `SuccessConnected` and `SuccessMultipleAccessPoints(Vec<...>)` since having an empty vector for "connected successfully" would make the API kind of awkward to use.

Reply:
You can always do `#[allow(clippy::large_enum_variant)]`.

With that said, and since you are anyway using allocations, I don't see the point of using `heapless::Vec` in the first place. Just use a regular `alloc::vec::Vec` then.

This would also make `Vec::clone` cheaper, of course.